templates/tftmpl/source_input.go

```diff
@@ -0,0 +1,96 @@
+package tftmpl
+
+import (
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/hashicorp/consul-terraform-sync/logging"
+	"github.com/hashicorp/hcl/v2/hclwrite"
+)
+
+var (
+	_ SourceInput = (*ServicesSourceInput)(nil)
+)
+
+// SourceInput handles appending a run source input's relevant templating for Terraform
+// generated files
+type SourceInput interface {
+	// ServicesAppended returns true if the services variable has been appended
+	// to the template content.
+	ServicesAppended() bool
+
+	// appendModuleAttribute writes to an HCL module body the SourceInput variable
+	// as a module argument in main.tf file.
+	// module "name" {
+	//   services = var.services
+	// }
+	appendModuleAttribute(*hclwrite.Body)
+
+	// appendTemplate writes the generated variable template for the SourceInput
+	// based on whether the source includes the source variable.
+	appendTemplate(io.Writer) error
+
+	// appendVariable writes the corresponding Terraform variable block to
+	// the variables.tf file.
+	appendVariable(io.Writer) error
+}
+
+// ServicesSourceInput handles appending a run source variable's relevant templating for Terraform
+// generated files
+type ServicesSourceInput struct {
+	Regexp string
+}
+
+// ServicesAppended returns true if the Regexp member variable is not an empty string
+func (sv ServicesSourceInput) ServicesAppended() bool {
+	return sv.Regexp != ""
+}
+
+func (sv ServicesSourceInput) appendModuleAttribute(*hclwrite.Body) {}
+
+func (sv ServicesSourceInput) appendTemplate(w io.Writer) error {
+	if sv.Regexp == "" {
+		return nil
+	}
+
+	// create the hcat query and write the template
+	q := sv.hcatQuery()
+	_, err := fmt.Fprintf(w, servicesRegexSourceInputIncludesVarTmpl, q)
+
+	if err != nil {
+		logging.Global().Named(logSystemName).Named(tftmplSubsystemName).Error(
+			"unable to write services source_input template", "error", err)
+		return err
+	}
+	return nil
+}
+
+func (sv ServicesSourceInput) appendVariable(io.Writer) error {
+	return nil
+}
+
+func (sv ServicesSourceInput) hcatQuery() string {
+	var opts []string
+
+	if sv.Regexp != "" {
+		opts = append(opts, fmt.Sprintf("regexp=%s", sv.Regexp))
+	}
+
+	if len(opts) > 0 {
+		return `"` + strings.Join(opts, `" "`) + `" ` // deliberate space at end
+	}
+	return ""
+}
+
+var servicesRegexSourceInputIncludesVarTmpl = `
+services = {
+{{- with $srv := servicesRegex %s }}
+  {{- range $s := $srv}}
+  "{{ joinStrings "." .ID .Node .Namespace .NodeDatacenter }}" = {
+{{ HCLService $s | indent 4 }}
+  },
+  {{- end}}
+{{- end}}
+}
+`
```

Comment:
Could you reuse variables when possible? There is a [variable that does this exact template already](https://github.com/hashicorp/consul-terraform-sync/blob/master/templates/tftmpl/condition.go#L104), and it would be easier to update the core template in one place if needed.